File: src/constants/goals.ts
import { GoalOption, PaceOption, GoalPace, Goal } from '../types';

export const GOAL_OPTIONS: GoalOption[] = [
  { value: 'fat_loss',             icon: '🔥', label: 'Lose Weight',           description: 'Burn fat through a calorie deficit' },
  { value: 'muscle_gain',          icon: '💪', label: 'Build Muscle',           description: 'Gain size and mass with a calorie surplus' },
  { value: 'body_recomp',          icon: '⚖️', label: 'Body Recomposition',     description: 'Lose fat and build muscle simultaneously' },
  { value: 'strength',             icon: '🏋️', label: 'Build Strength',         description: 'Increase your 1-rep maxes and raw power' },
  { value: 'toning',               icon: '✨', label: 'Tone & Define',           description: 'Lean out and sculpt visible definition' },
  { value: 'endurance',            icon: '🏃', label: 'Improve Endurance',      description: 'Run longer, recover faster, build stamina' },
  { value: 'athletic_performance', icon: '⚡', label: 'Athletic Performance',   description: 'Speed, power, and sport-specific fitness' },
  { value: 'maintain',             icon: '🎯', label: 'Maintain & Stay Active', description: 'Keep your fitness level and stay healthy' },
  { value: 'flexibility',          icon: '🧘', label: 'Flexibility & Mobility', description: 'Improve range of motion, reduce injury risk' },
  { value: 'stress_relief',        icon: '🌿', label: 'Mental Wellness',        description: 'Use movement to manage stress and mood' },
];

// Goals that show target weight + pace
export const WEIGHT_GOALS = new Set<Goal>(['fat_loss', 'toning', 'muscle_gain']);
// Goals that show timeline cards only
export const TIMELINE_GOALS = new Set<Goal>(['body_recomp', 'strength', 'endurance', 'athletic_performance']);
// Goals with no details step
export const LIFESTYLE_GOALS = new Set<Goal>(['maintain', 'flexibility', 'stress_relief']);

export const PACE_OPTIONS: Record<string, PaceOption[]> = {
  fat_loss: [
    { value: 'conservative', icon: '🐢', label: 'Slow & Steady', rate: '~0.5 lbs/week',  description: 'Sustainable, preserves muscle mass' },
    { value: 'moderate',     icon: '🚶', label: 'Balanced',      rate: '~1 lb/week',    description: 'Recommended for most people' },
    { value: 'aggressive',   icon: '🚀', label: 'Aggressive',    rate: '~1.5 lbs/week', description: 'Fastest results, more discipline needed' },
  ],
  toning: [
    { value: 'conservative', icon: '🐢', label: 'Gentle Cut',    rate: '~0.5 lbs/week',  description: 'Slow lean-out, very sustainable' },
    { value: 'moderate',     icon: '🚶', label: 'Steady Cut',    rate: '~0.75 lbs/week', description: 'Good balance of speed and comfort' },
    { value: 'aggressive',   icon: '🚀', label: 'Fast Cut',      rate: '~1 lb/week',     description: 'Quicker definition, stricter diet' },
  ],
  muscle_gain: [
    { value: 'conservative', icon: '🌱', label: 'Lean Bulk',       rate: '~0.25 lbs/week', description: 'Minimal fat gain, slow and clean' },
    { value: 'moderate',     icon: '💪', label: 'Standard Bulk',   rate: '~0.5 lbs/week',  description: 'Best balance of muscle vs fat' },
    { value: 'aggressive',   icon: '🦣', label: 'Aggressive Bulk', rate: '~1 lb/week',     description: 'Maximum muscle, expect some fat' },
  ],
  body_recomp: [
    { value: 'conservative', icon: '📅', label: '3 Months', rate: '12 weeks', description: 'Short commitment, build the habit' },
    { value: 'moderate',     icon: '🗓️', label: '6 Months', rate: '24 weeks', description: 'See significant visual changes' },
    { value: 'aggressive',   icon: '🏆', label: '1 Year',   rate: '52 weeks', description: 'Full body transformation' },
  ],
  strength: [
    { value: 'conservative', icon: '📅', label: '4 Weeks',  rate: '1 cycle',   description: 'Intro program to build base strength' },
    { value: 'moderate',     icon: '🗓️', label: '12 Weeks', rate: '1 program', description: 'Standard powerlifting block' },
    { value: 'aggressive',   icon: '🏆', label: '6 Months', rate: '2 blocks',  description: 'Long-term strength development' },
  ],
  endurance: [
    { value: 'conservative', icon: '📅', label: '4 Weeks',  rate: 'Intro',      description: 'Build aerobic base and consistency' },
    { value: 'moderate',     icon: '🗓️', label: '8 Weeks',  rate: 'Develop',    description: 'Noticeable improvement in stamina' },
    { value: 'aggressive',   icon: '🏆', label: '16 Weeks', rate: 'Race-ready', description: 'Train for a 10K or half marathon' },
  ],
  athletic_performance: [
    { value: 'conservative', icon: '📅', label: '4 Weeks',  rate: 'Foundation',  description: 'Movement quality and injury prevention' },
    { value: 'moderate',     icon: '🗓️', label: '12 Weeks', rate: 'Performance', description: 'Measurable sport-specific gains' },
    { value: 'aggressive',   icon: '🏆', label: '6 Months', rate: 'Elite',       description: 'Peak athletic conditioning' },
  ],
};

export const TIMELINE_WEEKS: Record<Goal, Record<GoalPace, number>> = {
  body_recomp:          { conservative: 12, moderate: 24, aggressive: 52 },
  strength:             { conservative: 4,  moderate: 12, aggressive: 26 },
  endurance:            { conservative: 4,  moderate: 8,  aggressive: 16 },
  athletic_performance: { conservative: 4,  moderate: 12, aggressive: 26 },
  fat_loss:             { conservative: 0,  moderate: 0,  aggressive: 0  },
  toning:               { conservative: 0,  moderate: 0,  aggressive: 0  },
  muscle_gain:          { conservative: 0,  moderate: 0,  aggressive: 0  },
  maintain:             { conservative: 0,  moderate: 0,  aggressive: 0  },
  flexibility:          { conservative: 0,  moderate: 0,  aggressive: 0  },
  stress_relief:        { conservative: 0,  moderate: 0,  aggressive: 0  },
};

export const GOAL_LABEL: Record<Goal, string> = Object.fromEntries(
  GOAL_OPTIONS.map(g => [g.value, g.label])
) as Record<Goal, string>;

File: backend/app/services/workout/goals.py
"""Single source of truth for workout-goal handling.

Every user-facing goal id resolves to exactly one canonical planner
bucket through `resolve_goal`. Callers that used to run their own
substring matches or keyword dictionaries now go through this module so
"which goals does the planner actually support?" has a concrete answer
backed by tests.

Supported set after this pass (see `supported_goal_ids`):
    fat_loss, muscle_gain, body_recomp, strength, endurance,
    athletic_performance, toning (alias → fat_loss),
    maintain (alias → body_recomp).

Dropped from the exposed set:
    flexibility, stress_relief — these require a mobility / low-
    intensity library the strength planner does not own, so exposing
    them as unique goals was misleading. They stay in the registry as
    `supported=False` so stale references (existing user profiles,
    telemetry) still resolve to a sensible fallback bucket.

Aliases
-------
An alias goal shares the same planner bucket as its canonical target.
`toning` normalizes to `fat_loss`, `maintain` normalizes to
`body_recomp`. The UI can still show them as distinct labels; the
planner treats them identically at the code level. This is the honest
middle-ground between "hide them" and "pretend they're different".
"""
from __future__ import annotations

from dataclasses import dataclass, field
from typing import Literal, Optional


PlannerBucket = Literal[
    "fat_loss",
    "muscle_gain",
    "body_recomp",
    "strength",
    "endurance",
    "athletic_performance",
    "hyrox",
    "general_health",
]


@dataclass(frozen=True)
class GoalDefinition:
    """One row in the registry.

    user_id     — the id the frontend sends on `PlanRequest.goal`.
    bucket      — the canonical planner bucket used by every downstream
                  decision (split, prescription, volume, recommendation).
    label       — user-facing display name.
    icon        — emoji shown in the goal picker.
    description — one-line marketing copy.
    supported   — true if this goal should appear in the UI goal picker.
                  Unsupported goals are still resolvable (so older user
                  profiles don't crash) but flagged for frontend hiding.
    alias_of    — if set, this goal shares its canonical behavior with
                  `alias_of`. Useful for UI variants that the planner
                  doesn't differentiate ("toning" ≈ "fat_loss").
    notes       — free-form explanation for the audit table.
    """
    user_id: str
    bucket: PlannerBucket
    label: str = ""
    icon: str = ""
    description: str = ""
    supported: bool = True
    alias_of: Optional[str] = None
    notes: str = ""


# Ordering matters for `supported_goal_ids()` — it's the display order
# used when seeding GoalOption rows.
_GOAL_REGISTRY: list[GoalDefinition] = [
    GoalDefinition(
        user_id="fat_loss",
        bucket="fat_loss",
        label="Lose Weight",
        icon="flame-outline",
        description="Burn fat through a sustainable calorie deficit",
        notes="Higher-rep circuits, shorter rest, conditioning-friendly.",
    ),
    GoalDefinition(
        user_id="muscle_gain",
        bucket="muscle_gain",
        label="Build Muscle",
        icon="barbell-outline",
        description="Gain size and strength with a calorie surplus",
        notes="Hypertrophy rep ranges, PPL/UL splits, 4 sets on primaries.",
    ),
    GoalDefinition(
        user_id="body_recomp",
        bucket="body_recomp",
        label="Body Recomposition",
        icon="swap-horizontal-outline",
        description="Lose fat and build muscle at the same time",
        notes="Moderate reps, balanced volume, works across day counts.",
    ),
    GoalDefinition(
        user_id="strength",
        bucket="strength",
        label="Build Strength",
        icon="trophy-outline",
        description="Increase your 1-rep maxes and raw power output",
        notes="Low-rep heavy compounds, longer rest, PPL-UL hybrid at 5+ days.",
    ),
    GoalDefinition(
        user_id="endurance",
        bucket="endurance",
        label="Improve Endurance",
        icon="heart-outline",
        description="Run longer, recover faster, and build stamina",
        notes="Cardio-first plans. Strength maintenance one day/week at 3+ days.",
    ),
    GoalDefinition(
        user_id="athletic_performance",
        bucket="athletic_performance",
        label="Athletic Performance",
        icon="flash-outline",
        description="Speed, power, agility, and sport-specific fitness",
        notes="Hybrid — alternates strength and conditioning days.",
    ),
    GoalDefinition(
        user_id="hyrox",
        bucket="hyrox",
        label="HYROX / Hybrid Race",
        icon="medal-outline",
        description="Train for HYROX, Deka, or hybrid fitness races",
        notes=(
            "Hybrid race prep: running, intervals, functional stations "
            "(sled push/pull, wall balls, lunges, row, ski erg, burpees). "
            "Conditioning-heavy with strength support, not bodybuilding."
        ),
    ),
    GoalDefinition(
        user_id="toning",
        bucket="fat_loss",
        label="Tone & Define",
        icon="sparkles-outline",
        description="Lean out and sculpt visible muscle definition",
        alias_of="fat_loss",
        notes="UI variant of fat_loss. Same planner behavior.",
    ),
    GoalDefinition(
        user_id="maintain",
        bucket="general_health",
        label="Maintain & Stay Active",
        icon="checkmark-circle-outline",
        description="Keep your current fitness level and stay healthy",
        alias_of="general_health",
        notes=(
            "Routed to the general_health profile which mixes one lift "
            "day, one easy cardio day, and one mobility day per week."
        ),
    ),
    # Hidden from UI — the planner cannot honestly differentiate these
    # without a mobility / yoga exercise library it does not own. The
    # registry still resolves them (to general_health) so older user
    # profiles don't crash on plan generation.
    GoalDefinition(
        user_id="flexibility",
        bucket="general_health",
        label="Flexibility & Mobility",
        icon="body-outline",
        description="Improve range of motion and reduce injury risk",
        supported=False,
        notes="Needs a mobility library the strength planner doesn't own.",
    ),
    GoalDefinition(
        user_id="stress_relief",
        bucket="general_health",
        label="Mental Wellness",
        icon="leaf-outline",
        description="Use movement to manage stress and improve mood",
        supported=False,
        notes="Needs a low-intensity / yoga library the planner doesn't own.",
    ),
]


_BY_ID: dict[str, GoalDefinition] = {g.user_id: g for g in _GOAL_REGISTRY}


# Legacy and long-form goal ids from goalConfig.ts that aren't first-
# class registry entries but should still resolve to the right bucket.
# Kept short and obvious. Add to the registry proper if any of these
# ever become first-class user-facing goals.
_LEGACY_ALIASES: dict[str, str] = {
    # Fat loss synonyms
    "lose_fat":              "fat_loss",
    "cut":                   "fat_loss",
    "get_lean":              "fat_loss",
    "preserve_muscle_cutting": "fat_loss",
    # Muscle gain synonyms
    "build_muscle":          "muscle_gain",
    "lean_bulk":             "muscle_gain",
    "bulk":                  "muscle_gain",
    "gain_weight":           "muscle_gain",
    "improve_aesthetics":    "muscle_gain",
    "build_glutes":          "muscle_gain",
    "hypertrophy":           "muscle_gain",
    # Strength synonyms
    "build_strength":        "strength",
    "powerlifting":          "strength",
    "strength_training":     "strength",
    "get_stronger":          "strength",
    "improve_1rm":           "strength",
    "improve_bench":         "strength",
    "improve_deadlift":      "strength",
    "improve_squat":         "strength",
    "functional_strength":   "strength",
    "relative_strength":     "strength",
    # Endurance synonyms — goalConfig.ts has a long list under
    # `cardio_endurance`; all of them route here.
    "cardio_endurance":      "endurance",
    "cardio":                "endurance",
    "improve_cardio":        "endurance",
    "improve_conditioning":  "endurance",
    "aerobic_base":          "endurance",
    "improve_vo2":           "endurance",
    "increase_stamina":      "endurance",
    "running_fitness":       "endurance",
    "train_5k":              "endurance",
    "train_10k":             "endurance",
    "train_half":            "endurance",
    "train_marathon":        "endurance",
    "sprint_speed":          "endurance",
    "interval_perf":         "endurance",
    "run_faster":            "endurance",
    "marathon":              "endurance",
    # Athletic synonyms
    "sport_performance":     "athletic_performance",
    "improve_athleticism":   "athletic_performance",
    # Body recomp synonyms
    "body_recomposition":    "body_recomp",
    "recomp":                "body_recomp",
    # HYROX / hybrid race synonyms
    "hybrid_race":           "hyrox",
    "race_hybrid":           "hyrox",
    "deka":                  "hyrox",
    "deka_fit":              "hyrox",
    "functional_fitness":    "hyrox",
}


def resolve_goal(goal: Optional[str]) -> GoalDefinition:
    """Look up a goal id in the registry with alias + substring fallback.

    Order of resolution:
      1. Exact match against `_BY_ID` (the first-class registry).
      2. Exact match against `_LEGACY_ALIASES` → redirected to the
         canonical registry entry.
      3. Substring match on obvious keywords (cardio/run/endurance,
         athletic/sport, muscle/hypertrophy, strength, fat/lose/cut).
      4. Final fallback: `body_recomp` (our "no strong opinion"
         default — NOT `general_health`, because that bucket is
         reserved for hidden unsupported goals).

    Never returns `None` — planner code can always assume a valid
    definition."""
    if not goal:
        return _BY_ID["body_recomp"]
    g = str(goal).strip().lower()
    if g in _BY_ID:
        return _BY_ID[g]
    canonical = _LEGACY_ALIASES.get(g)
    if canonical and canonical in _BY_ID:
        return _BY_ID[canonical]
    # Substring fallback for free-form goals that didn't land in either
    # table. Matches the keyword-scan the old `_goal_bucket` used to do.
    if "cardio" in g or "endurance" in g or "run" in g or "marathon" in g:
        return _BY_ID["endurance"]
    if "hyrox" in g or "hybrid race" in g or "deka" in g:
        return _BY_ID["hyrox"]
    if "athletic" in g or "sport" in g:
        return _BY_ID["athletic_performance"]
    if "muscle" in g or "hypertrophy" in g or "bulk" in g:
        return _BY_ID["muscle_gain"]
    if "strength" in g or "powerlift" in g:
        return _BY_ID["strength"]
    if "fat" in g or "lose" in g or "cut" in g or "lean" in g:
        return _BY_ID["fat_loss"]
    return _BY_ID["body_recomp"]


def goal_bucket(goal: Optional[str]) -> str:
    """Canonical planner bucket for `goal`. Single source of truth —
    every planner branch should call this, never roll its own mapping."""
    return resolve_goal(goal).bucket


def is_goal_supported(goal: Optional[str]) -> bool:
    """Whether `goal` should appear in the UI goal picker. Unsupported
    goals still resolve (for stale profiles), they just shouldn't be
    offered to new users."""
    return resolve_goal(goal).supported


def supported_goal_ids() -> list[str]:
    """Ordered list of user-facing goal ids the UI should show.

    Used by `seed.py` to populate `GoalOption` rows and by tests to
    verify every exposed goal has intentional planner behavior."""
    return [g.user_id for g in _GOAL_REGISTRY if g.supported]


def goal_registry_rows() -> list[GoalDefinition]:
    """Read-only view of the registry for audit tables / tests."""
    return list(_GOAL_REGISTRY)


def ui_goal_rows() -> list[tuple[str, str, str, str]]:
    """Return `(user_id, label, icon, description)` tuples for every
    supported goal in display order. Used by `seed.py` to populate
    `GoalOption` rows — keeps the UI metadata and the planner registry
    in a single place so they can't drift."""
    return [
        (g.user_id, g.label, g.icon, g.description)
        for g in _GOAL_REGISTRY
        if g.supported
    ]
